test/common/http/http2/http2_connection_fuzz_test.cc

```diff
@@ -0,0 +1,314 @@
+#include <alloca.h>
+
+#include <string>
+#include <vector>
+
+#include "source/common/http/http2/codec_impl.h"
+
+#include "test/common/http/http2/http2_connection.pb.h"
+#include "test/fuzz/fuzz_runner.h"
+#include "test/mocks/http/mocks.h"
+#include "test/mocks/network/mocks.h"
+#include "test/test_common/test_runtime.h"
+
+#include "absl/strings/string_view.h"
+#include "nghttp2/nghttp2.h"
+
+namespace Envoy {
+namespace Http {
+namespace Http2 {
+namespace {
+
+static const uint32_t MAX_HEADERS = 32;
+static const size_t MAX_HD_TABLE_BUF_SIZE = 4096;
+
+class Http2Frame {
+public:
+  static constexpr char Preamble[25] = "PRI * HTTP/2.0\r\n\r\nSM\r\n\r\n";
+  static constexpr size_t HeaderSize = 9;
+  Http2Frame(uint32_t stream_id, uint8_t type, uint8_t flags) : stream_id_(stream_id) {
+    data_.assign(HeaderSize, 0);
+    setType(type);
+    setFlags(flags);
+  }
+
+  Http2Frame(absl::string_view contents) : stream_id_(1) {
+    appendData(contents);
+    if (data_.size() >= HeaderSize) {
+      std::memcpy(&stream_id_, &data_[5], sizeof(stream_id_));
+    }
+  }
+
+  void setPayloadSize(uint32_t size) {
+    data_[0] = (size >> 16) & 0xff;
+    data_[1] = (size >> 8) & 0xff;
+    data_[2] = size & 0xff;
+  }
+
+  void adjustPayloadSize() { setPayloadSize(data_.size() - HeaderSize); }
+
+  void setType(uint8_t type) { data_[3] = type; }
+  void setFlags(uint8_t flags) { data_[4] = flags; }
+
+  void makeClientStreamId() {
+    uint32_t id = (stream_id_ << 1) + 1;
+    if (data_.size() >= HeaderSize) {
+      std::memcpy(&data_[5], &id, sizeof(id));
+    }
+  }
+
+  void makeServerStreamId() {
+    uint32_t id = stream_id_ << 1;
+    if (data_.size() >= HeaderSize) {
+      std::memcpy(&data_[5], &id, sizeof(id));
+    }
+  }
+
+  void appendData(absl::string_view data) { data_.insert(data_.end(), data.begin(), data.end()); }
+
+  const absl::string_view payload() const {
+    const char* d = reinterpret_cast<const char*>(data_.data());
+    return absl::string_view(d, data_.size());
```

Comment:
Shouldn't the return value omit the frame header?

Reply:
I renamed the method to 'serialize' to better reflect that the method returns the wire format of the frame to feed to the decoder. The frame header needs to be included in the wire format.